libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,575 @@
+  var ctx: sha256
+  ctx.init()
+  ctx.update(ss.h.data)
+  ctx.update(data)
+  ss.h = ctx.finish()
+  trace "mixHash", hash = ss.h.data
+
+proc mixKeyAndHash(ss: var SymmetricState; ikm: var openarray[byte]) =
+  var
+    temp_keys: array[3, ChaChaPolyKey]
+  sha256.hkdf(ss.ck, ikm, [], temp_keys)
+  ss.ck = temp_keys[0]
+  ss.mixHash(temp_keys[1])
+  init ss.cs, temp_keys[2]
+
+proc encryptAndHash(ss: var SymmetricState, data: var openarray[byte]): seq[byte] =
+  # according to spec if key is empty leave plaintext
+  if ss.cs.hasKey:
+    result = ss.cs.encryptWithAd(ss.h.data, data)
+  else:
+    result = @data
+  ss.mixHash(result)
+
+proc decryptAndHash(ss: var SymmetricState, data: var openarray[byte]): seq[byte] =
+  # according to spec if key is empty leave plaintext
+  if ss.cs.hasKey:
+    result = ss.cs.decryptWithAd(ss.h.data, data)
+  else:
+    result = @data
+  ss.mixHash(data)
+
+proc split(ss: var SymmetricState): tuple[cs1, cs2: CipherState] =
+  var
+    temp_keys: array[2, ChaChaPolyKey]
+  sha256.hkdf(ss.ck, [], [], temp_keys)
+  init result.cs1, temp_keys[0]
+  init result.cs2, temp_keys[1]
+  
+template write_e: untyped =
+  trace "noise write e"
+  # Sets e (which must be empty) to GENERATE_KEYPAIR(). Appends e.public_key to the buffer. Calls MixHash(e.public_key).
+  hs.e = genKeyPair()
+  msg &= hs.e.publicKey
+  hs.ss.mixHash(hs.e.publicKey)
+
+template write_s: untyped =
+  trace "noise write s"
+  # Appends EncryptAndHash(s.public_key) to the buffer.
+  msg &= hs.ss.encryptAndHash(hs.s.publicKey)
+
+template dh_ee: untyped =
+  trace "noise dh ee"
+  # Calls MixKey(DH(e, re)).
+  hs.ss.mixKey(dh(hs.e.privateKey, hs.re))
+
+template dh_es: untyped =
+  trace "noise dh es"
+  # Calls MixKey(DH(e, rs)) if initiator, MixKey(DH(s, re)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+  else:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+
+template dh_se: untyped =
+  trace "noise dh se"
+  # Calls MixKey(DH(s, re)) if initiator, MixKey(DH(e, rs)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+  else:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+
+template dh_ss: untyped =
+  trace "noise dh ss"
+  # Calls MixKey(DH(s, rs)).
+  hs.ss.mixKey(dh(hs.s.privateKey, hs.rs))
+
+template read_e: untyped =
+  trace "noise read e", size = msg.len
+
+  if msg.len < Curve25519Key.len:
+    raise newException(NoiseHandshakeError, "Noise E, expected more data")
+
+  # Sets re (which must be empty) to the next DHLEN bytes from the message. Calls MixHash(re.public_key).
+  copyMem(addr hs.re[0], addr msg[0], Curve25519Key.len)
+  msg = msg[Curve25519Key.len..msg.high]
+  hs.ss.mixHash(hs.re)
+
+template read_s: untyped =
+  trace "noise read s", size = msg.len
+  # Sets temp to the next DHLEN + 16 bytes of the message if HasKey() == True, or to the next DHLEN bytes otherwise.
+  # Sets rs (which must be empty) to DecryptAndHash(temp).
+  var temp: seq[byte]
```

Comment:
this can be written in a safer way:

```
func popPrefix(v: var seq, len: int): seq = ..
let tmp = if bla.haskey:
   if msg.len < 16: ...
   popPrefix(msg, 16)
  else: popPrefix(msg, ..)
```
where popPrefix does the prefix cut - more elegant in that there's less var mutation restricting the mistakes that can happen - this comment applies throughout the code, which is very `var`-heavy unnecessarily

Reply:
I didn't get this one, 
You don't like `arr[0..X]` slicing?
Or mostly the reason is cos actually it's creating a new seq all the time? (that's a fair point indeed, but slicing is nice imho)